Bring stacking under control (#1387)
silx-kit/h5web: apps/storybook/src/Stacking.stories.mdx

Some of these elements must appear above or behind others, so it's important to control and understand their stacking order.

The closest container to all these elements, `canvasWrapper`, creates a new [stacking context](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context).
This allows its children (both direct and nested) to be stacked in relation with one another within this stacking context (unless some of them define their own stacking context).

By default, the stacking order is as follows, from furthest (behind) to closest (in front):

1. axis system (including the axis grid);
2. SVG elements rendered via [`SvgElement`](http://localhost:6006/?path=/story/building-blocks-svgelement--default) into the `svg` overlay, in DOM order;
3. HTML elements rendered via [`Html`](http://localhost:6006/?path=/story/building-blocks-html--default), in DOM order (including [`Annotation`](http://localhost:6006/?path=/story/building-blocks-annotation--default) and [`Overlay`](http://localhost:6006/?path=/story/building-blocks-overlay--default));
4. tooltip

> 5. floating toolbar (typically with reset zoom button)

Author: I've added a schema and emphasised/reworded a few things (in a separate commit to ease your review).

  in front of other elements that remained in the DOM. This can be worked around in a few ways, as explained below.

### How to avoid unmounting `Html` and `SvgElement`

The simplest solution is to move the rendering condition to the children. This is also better for performance, as mounting `Html` and `SvgElement` is costly.

```tsx
<>
  {/* Avoid */}
  {bool && (
    <Annotation x={0} y={0}>
      Foo
    </Annotation>
  )}
  {bool && (
    <Html>
      <p>Bar</p>
    </Html>
  )}
  {bool && (
    <SvgElement>
      <SvgRect />
    </SvgElement>
  )}

  {/* Prefer */}
  <Annotation x={0} y={0}>
    {bool && 'Foo'}
  </Annotation>
  <Html>{bool && <p>Bar</p>}</Html>
  <SvgElement>{bool && <SvgRect />}</SvgElement>
</>
```

Another great solution is to simply hide the children conditionally:

```tsx
<>
  <Html x={0} y={0}>
    <span hidden={bool || undefined}>Foo</span>
  </Html>

  {/* Hide `Annotation`'s internal `div` (works with `Overlay` as well) */}
  <Annotation x={0} y={0} hidden={bool || undefined}>
    Foo
  </Annotation>

  <SvgElement>
    <SvgRect fill={bool ? 'skyblue' : 'transparent'} />
  </SvgElement>
</>
```

> ### What if DOM order cannot be changed or guaranteed?

Author: I've reworded the first sentence of this section to:

> If you can't restructure your React tree to stack HTML and SVG elements in the right order,
or if you can't avoid unmounting/remounting some occurrences of `Html` and `SvgElement`,
the last resort is to control the stacking order explicitly with `z-index`.

For `Html`, the last resort is to control the stacking order explicitly with `z-index`. Knowing that, by default, `Html`'s internal `div` element has no `z-index`,
which is equivalent to `z-index: 0`, you can move an `Html` element in front of, or behind, other `Html` elements simply by giving it a positive or negative
`z-index` (respectively).

In the following example, the stacking order is maintained regardless of the React rendering order:

```tsx
<>
  {foo && <Annotation style={{ zIndex: 1 }}>Front</Annotation>}
  {bar && <Annotation style={{ zIndex: -1 }}>Back</Annotation>}
  <Annotation>Middle</Annotation>
</>
```

> To avoid messing with the stacking order of other components (grid, tooltip, etc.), please remain within the range [-999, 999].

For `SvgElement`, the problem is more complicated as SVG 1.1 does not provide an equivalent to `z-index`.
A workaround is to not use `SvgElement` at all and instead render an `svg` element yourself via `Html`, with all the caveats
that this entails:

```tsx
<Html>
  <svg style={{ zIndex: 1 }}>{/* ... */}</svg>
</Html>
```

### Crossing stacking boundaries

> 

Author: I've reworded the section.

### How to avoid unmounting `Html` and `SvgElement`

This is also better for performance, as mounting `Html` and `SvgElement` is costly. The simplest solution is to move the rendering condition to the children:

```tsx
<>
  {/* Avoid */}
  {bool && (
    <Annotation x={0} y={0}>
      Foo
    </Annotation>
  )}
  {bool && (
    <Html>
      <p>Bar</p>
    </Html>
  )}
  {bool && (
    <SvgElement>
      <SvgRect />
    </SvgElement>
  )}

  {/* Prefer */}
  <Annotation x={0} y={0}>
    {bool && 'Foo'}
  </Annotation>
  <Html>{bool && <p>Bar</p>}</Html>
  <SvgElement>{bool && <SvgRect />}</SvgElement>
</>
```

Another solution is to conditionally hide the children:

```tsx
<>
  <Html x={0} y={0}>
    <span hidden={bool || undefined}>Foo</span>
  </Html>

  {/* Hide `Annotation`'s internal `div` (works with `Overlay` as well) */}
  <Annotation x={0} y={0} hidden={bool || undefined}>
    Foo
  </Annotation>

  <SvgElement>
    <SvgRect fill={bool ? 'skyblue' : 'transparent'} />
  </SvgElement>
</>
```

### What if DOM order of HTML elements cannot be changed or guaranteed?

> If you can't restructure your React tree to render `Html` elements in the order you want them to be stacked,

Reviewer: ```suggestion
If you can't restructure your React tree to render `Html` elements in an order following the desired stacking,
```

or if you can't avoid unmounting/remounting them, the last resort is to control the stacking order explicitly with `z-index`.

> Knowing that, by default, `Html`'s internal `div` element has no `z-index`,
> which is equivalent to `z-index: 0`, you can move an `Html` element in front of,
> or behind, other `Html` elements simply by giving it a positive or negative
> `z-index` (respectively).

Reviewer: Lots of info in this: I suggest to break it down in two sentences.

```suggestion
By default, `Html`'s internal `div` element has no `z-index`,
this is equivalent to `z-index: 0`. 
You can therefore move an `Html` element in front of,
or behind, other `Html` elements simply by giving it a positive or negative
`z-index` (respectively).
```

Author: Good call 👍

In the following example, the stacking order is maintained regardless of the React rendering order:

```tsx
<>
  {foo && <Annotation style={{ zIndex: 1 }}>Front</Annotation>}
  {bar && <Annotation style={{ zIndex: -1 }}>Back</Annotation>}
  <Annotation>Middle</Annotation>
</>
```

> To avoid messing with the stacking order of other components (grid, tooltip, etc.), please remain within the range [-999, 999].

### What if DOM order of SVG elements cannot be changed or guaranteed?

For `SvgElement`, the problem is more complicated as SVG 1.1 does not provide an equivalent to `z-index`.
A workaround is to not use `SvgElement` at all and instead render an `svg` element yourself via `Html`, with the caveats

> that this entails (i.e. you have to position and size the SVG element yourself):

Reviewer: :+1:

1. **axis system** (including the axis grid);
2. **SVG elements** rendered via [`SvgElement`](http://localhost:6006/?path=/story/building-blocks-svgelement--default) into the `svg` overlay, in DOM order;
3. **HTML elements** rendered via [`Html`](http://localhost:6006/?path=/story/building-blocks-html--default), in DOM order (including [`Annotation`](http://localhost:6006/?path=/story/building-blocks-annotation--default) and [`Overlay`](http://localhost:6006/?path=/story/building-blocks-overlay--default));
4. **tooltip**
5. **floating toolbar** (typically with reset zoom button)

```
┌──────────────────────────────────┐
│ 1. Axis system                   │
│   ┌──────────────────────────────┴───┐
│   │ 2. SVG elements                  │
└── ┤   ┌──────────────────────────────┴───┐
    │   │ 3. HTML elements                 │
    └── ┤   ┌──────────────────────────────┴───┐
        │   │ 4. Tooltip                       │
        └── ┤   ┌──────────────────────────────┴───┐
            │   │ 5. Floating toolbar              │
            └── ┤                                  │
                │                                  │

>                 └──────────────────────────────────┘

Reviewer: That's so nice :joy:

Author: Took me age 😂
